File: iOS/PrivateVPNPacketTunnel/WGRelayClient.swift
import Darwin
import Foundation
import Network
import os

/// WireGuard-over-TCP relay client — the iOS counterpart of the Android `WGRelay`.
///
/// The WireGuard interface inside this extension is pointed at a local UDP listener
/// here. Every datagram it sends is wrapped in a length-prefixed TCP frame
/// (`[length: u16 big-endian][payload]`) and shipped to the relay daemon running next
/// to the exit node, which unwraps it and forwards it to the WireGuard server's UDP
/// port (and back). This is how the tunnel survives networks where the WireGuard
/// handshake never completes over raw UDP.
///
/// The local UDP port stays the same for the whole tunnel lifetime, including across
/// TCP reconnects, so WireGuard never has to be reconfigured: it keeps sending
/// handshakes into the local listener and the relay re-establishes the link.
final class WGRelayClient {

    private let host: String
    /// Relay ports to try, in order (node-1 runs 9444 + 8443, node-2 only 8443).
    private let ports: [UInt16]
    private let log: Logger
    private let listener: RelayUDPListener
    private var portIndex = 0
    private var lastGoodPort: UInt16?

    private let lock = NSLock()
    private var tcpFD: Int32 = -1
    private var running = false

    init(host: String, ports: [UInt16], log: Logger) {
        self.host = host
        self.ports = ports.isEmpty ? [9444] : ports
        self.log = log
        self.listener = RelayUDPListener(label: "relay-udp", log: log)
    }

    /// True while the TCP link to the relay is established. The tunnel uses this to
    /// decide whether the local listener is actually carrying anything.
    var isConnected: Bool {
        lock.lock()
        defer { lock.unlock() }
        return tcpFD >= 0
    }

    /// Opens the local UDP listener and starts the TCP link.
    /// - Returns: the local UDP port WireGuard must send its datagrams to.
    func start() throws -> UInt16 {
        let localPort = try listener.start { [weak self] datagram in
            self?.forwardToRelay(datagram)
        }

        lock.lock()
        running = true
        lock.unlock()

        Thread { [weak self] in
            while let self, self.running {
                Thread.sleep(forTimeInterval: 15)
                self.note("heartbeat \(self.countersSummary())")
            }
        }.start()

        // The TCP link is established (and re-established) on its own thread, so a
        // relay that is briefly unreachable never blocks the tunnel from starting.
        let readThread = Thread { [weak self] in self?.tcpReadLoop() }
        readThread.name = "wg-relay-tcp-read"
        readThread.start()

        note("local udp listener 127.0.0.1:\(localPort) -> \(self.host):\(self.ports)")
        return localPort
    }

    func stop() {
        lock.lock()
        guard running else { lock.unlock(); return }
        running = false
        let tcp = tcpFD
        tcpFD = -1
        lock.unlock()

        // Đóng socket UDP + CHỜ thread đọc thoát hẳn trước khi trả về, để phiên sau bind
        // lại được và không bị thread cũ của phiên này đọc nhầm fd (xem RelayUDPListener).
        listener.stop()
        // TCP: shutdown để đánh thức `recv` đang chặn; chính `tcpReadLoop` sở hữu fd nên
        // nó tự đóng, tránh đóng trùng.
        if tcp >= 0 { shutdown(tcp, SHUT_RDWR) }
        note("stopped")
    }

    // MARK: - TCP link

    /// Connects to the relay, retrying for as long as the tunnel is up.
    private func connectTCP() -> Int32? {
        while running {
            let fd = socket(AF_INET, SOCK_STREAM, 0)
            guard fd >= 0 else {
                log.error("relay: tcp socket errno=\(errno)")
                return nil
            }

            let port = lastGoodPort ?? ports[portIndex % ports.count]
            if lastGoodPort == nil { portIndex += 1 }

            var address = sockaddr_in()
            address.sin_family = sa_family_t(AF_INET)
            address.sin_port = port.bigEndian
            if inet_pton(AF_INET, host, &address.sin_addr) != 1 {
                // Hostname: resolve it (the relay host is normally an IP literal).
                var hints = addrinfo()
                hints.ai_family = AF_INET
                hints.ai_socktype = SOCK_STREAM
                var result: UnsafeMutablePointer<addrinfo>?
                if getaddrinfo(host, nil, &hints, &result) == 0, let info = result {
                    address.sin_addr = withUnsafePointer(to: info.pointee.ai_addr) {
                        $0.withMemoryRebound(to: sockaddr_in.self, capacity: 1) { $0.pointee.sin_addr }
                    }
                    freeaddrinfo(result)
                } else {
                    log.error("relay: cannot resolve \(self.host)")
                    close(fd)
                    Thread.sleep(forTimeInterval: 2)
                    continue
                }
            }

            // Non-blocking connect so a black-holing relay cannot freeze the thread.
            let flags = fcntl(fd, F_GETFL, 0)
            _ = fcntl(fd, F_SETFL, flags | O_NONBLOCK)
            let started = Date()
            let connected = withUnsafePointer(to: &address) { pointer in
                pointer.withMemoryRebound(to: sockaddr.self, capacity: 1) {
                    Darwin.connect(fd, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
                }
            }
            var ok = connected == 0
            // Lý do hỏng phải ghi RÕ, không dùng lại `errno`: khi connect không hoàn tất thì
            // `errno` vẫn là EINPROGRESS (36) — hằng số của chính lời gọi connect (Darwin),
            // hoàn toàn không phải lỗi. Log cũ in "failed errno=36" nên đọc lên tưởng
            // EINPROGRESS là nguyên nhân, trong khi thật ra là select hết hạn chờ. Đã mất
            // thời gian vì chỗ này khi debug "connected nhưng không có mạng" trên iPad.
            var failureReason = connected == 0 ? "" : "errno=\(errno)"
            if !ok && errno == EINPROGRESS {
                var writeSet = fd_set()
                setFD(fd, &writeSet)
                var timeout = timeval(tv_sec: 10, tv_usec: 0)
                let selected = select(fd + 1, nil, &writeSet, nil, &timeout)
                if selected > 0 {
                    var error: Int32 = 0
                    var errorLength = socklen_t(MemoryLayout<Int32>.size)
                    _ = getsockopt(fd, SOL_SOCKET, SO_ERROR, &error, &errorLength)
                    ok = error == 0
                    failureReason = error == 0 ? "" : "SO_ERROR=\(error)"
                } else if selected == 0 {
                    failureReason = "hết hạn 10s chờ connect hoàn tất (EINPROGRESS, không phải lỗi)"
                } else {
                    failureReason = "select errno=\(errno)"
                }
            }
            _ = fcntl(fd, F_SETFL, flags)

            guard ok else {
                note("connect \(self.host):\(port) failed: \(failureReason)")
                close(fd)
                Thread.sleep(forTimeInterval: 2)
                continue
            }

            var noDelay: Int32 = 1
            _ = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &noDelay, socklen_t(MemoryLayout<Int32>.size))
            lastGoodPort = port
            note("connected to \(self.host):\(port) in \(Int(Date().timeIntervalSince(started) * 1000))ms")
            return fd
        }
        return nil
    }

    /// Reads length-prefixed frames and hands each payload to WireGuard as a datagram.
    private func tcpReadLoop() {
        while running {
            guard let fd = connectTCP() else { return }
            lock.lock()
            tcpFD = fd
            lock.unlock()

            var header = [UInt8](repeating: 0, count: 2)
            while running {
                guard readExactly(fd, into: &header, count: 2) else { break }
                let frameLength = Int(header[0]) << 8 | Int(header[1])
                guard frameLength > 0, frameLength <= 65_535 else { continue }
                var payload = [UInt8](repeating: 0, count: frameLength)
                guard readExactly(fd, into: &payload, count: frameLength) else { break }
                receivedFrames += 1
                receivedBytes += frameLength
                listener.sendToPeer(payload)
            }

            lock.lock()
            if tcpFD == fd { tcpFD = -1 }
            lock.unlock()
            shutdown(fd, SHUT_RDWR)
            close(fd)
            if running { note("TCP link dropped, reconnecting") }
        }
    }

    /// Đẩy một datagram WireGuard nhận từ listener UDP vào link TCP, đóng gói
    /// `[length: u16 big-endian][payload]`. Chạy trên thread đọc của listener.
    private func forwardToRelay(_ datagram: Data) {
        lock.lock()
        let fd = tcpFD
        lock.unlock()

        // No link yet: drop it — WireGuard retries its handshake every 5s anyway.
        guard fd >= 0 else {
            lock.lock(); udpDroppedNoLink += 1; lock.unlock()
            return
        }

        var frame = [UInt8](repeating: 0, count: datagram.count + 2)
        frame[0] = UInt8((datagram.count >> 8) & 0xff)
        frame[1] = UInt8(datagram.count & 0xff)
        frame.replaceSubrange(2..<(datagram.count + 2), with: datagram)
        lock.lock()
        sentFrames += 1
        sentBytes += datagram.count
        lock.unlock()
        if !writeAll(fd, frame) {
            // The link is one-way broken (hotel NAT / relay restart): tear it down
            // so the read loop reconnects. Without this, every later packet was
            // dropped silently and the tunnel stayed "up" with no traffic.
            note("write to relay failed errno=\(errno) — dropping link to reconnect")
            dropLink(fd, reason: "write failure")
        }
    }

    // MARK: - Socket helpers

    // MARK: - Diagnostics

    private var sentFrames = 0
    private var sentBytes = 0
    private var receivedFrames = 0
    private var receivedBytes = 0
    private var udpDroppedNoLink = 0

    /// Logs to os_log at `.default` (visible without `--info`) *and* to the extension's
    /// diagnostic file. `privacy: .public` so the content is not redacted.
    private func note(_ message: String) {
        log.log(level: .default, "relay: \(message, privacy: .public)")
        RelayDiagnostics.shared.log("relay: \(message)")
    }

    /// Counters useful to spot which direction of the bridge went quiet.
    func countersSummary() -> String {
        lock.lock()
        defer { lock.unlock() }
        return "udpFrames=\(sentFrames) udpBytes=\(sentBytes) framesFromRelay=\(receivedFrames) bytesFromRelay=\(receivedBytes) droppedNoLink=\(udpDroppedNoLink) tcpFd=\(tcpFD)"
    }

    /// Closes a broken link; the read loop notices and reconnects.
    private func dropLink(_ fd: Int32, reason: String) {
        note("dropping TCP link (fd=\(fd)): \(reason)")
        lock.lock()
        if tcpFD == fd { tcpFD = -1 }
        lock.unlock()
        shutdown(fd, SHUT_RDWR)
        close(fd)
    }

    private func readExactly(_ fd: Int32, into buffer: inout [UInt8], count: Int) -> Bool {
        var offset = 0
        while offset < count {
            let read = buffer.withUnsafeMutableBytes { raw -> Int in
                guard let base = raw.baseAddress else { return -1 }
                return recv(fd, base + offset, count - offset, 0)
            }
            if read > 0 {
                offset += read
            } else if read == 0 {
                return false // peer closed
            } else if errno == EINTR {
                continue
            } else {
                return false
            }
        }
        return true
    }

    private func writeAll(_ fd: Int32, _ bytes: [UInt8]) -> Bool {
        var offset = 0
        while offset < bytes.count {
            let written = bytes.withUnsafeBytes { raw -> Int in
                guard let base = raw.baseAddress else { return -1 }
                return send(fd, base + offset, bytes.count - offset, 0)
            }
            if written > 0 {
                offset += written
            } else if written < 0 && errno == EINTR {
                continue
            } else {
                return false
            }
        }
        return true
    }

    private func setFD(_ fd: Int32, _ set: inout fd_set) {
        let index = Int(fd) / 32
        let bit = Int32(1) << (Int(fd) % 32)
        withUnsafeMutablePointer(to: &set.fds_bits) { pointer in
            pointer.withMemoryRebound(to: Int32.self, capacity: 32) { bits in
                bits[index] |= bit
            }
        }
    }
}

File: windows/PrivateVPNWindows.Core/Tunnel/WgRelayClient.cs
using System.Diagnostics;
using System.Net;
using System.Net.Sockets;

namespace VpnFlow.Core.Tunnel;

/// <summary>
/// WireGuard-over-TCP relay client — port trung thực của
/// <c>WGRelayClient.swift</c> (iOS/PrivateVPNPacketTunnel/WGRelayClient.swift:18-388).
///
/// WireGuard trỏ vào listener UDP local ở đây. Mỗi datagram được bọc trong frame
/// TCP độ dài tiền tố <c>[length: u16 big-endian][payload]</c> rồi gửi tới relay
/// daemon cạnh exit node; relay bung ra UDP tới cổng WireGuard của node (và trả lời
/// ngược lại). Cùng giao thức với <c>tools/node-setup/relay.go:59-102</c>.
///
/// Cổng UDP local giữ nguyên suốt vòng đời tunnel, kể cả khi TCP reconnect, nên
/// WireGuard không phải cấu hình lại.
/// </summary>
public sealed class WgRelayClient : ITunnelTransport
{
    private const int MaxDatagramBytes = 65_535;
    private const int ConnectTimeoutMs = 10_000;
    private const int RetryDelayMs = 2_000;
    private const int HeartbeatSeconds = 15;

    private readonly string _host;
    private readonly ushort[] _ports;
    private readonly ITunnelLogger _log;
    private readonly object _lock = new();
    private readonly CancellationTokenSource _cts = new();

    private int _portIndex;
    private ushort? _lastGoodPort;

    private Socket? _udp;
    private Socket? _tcp;
    private bool _running;
    private IPEndPoint? _peer;

    private Task? _readTask;
    private Task? _writeTask;
    private Task? _heartbeatTask;

    private int _sentFrames;
    private long _sentBytes;
    private int _receivedFrames;
    private long _receivedBytes;
    private int _udpDroppedNoLink;

    public WgRelayClient(string host, IReadOnlyList<ushort> ports, ITunnelLogger log, string? transportId = null)
    {
        _host = host ?? throw new ArgumentNullException(nameof(host));
        _ports = ports is { Count: > 0 } ? ports.ToArray() : new ushort[] { 9444 };
        _log = log ?? throw new ArgumentNullException(nameof(log));
        TransportId = transportId ?? "tcp";
    }

    public string TransportId { get; }

    public ushort LocalPort { get; private set; }

    /// <summary>True khi link TCP tới relay đã thiết lập.</summary>
    public bool IsConnected
    {
        get
        {
            lock (_lock)
            {
                return _tcp is not null;
            }
        }
    }

    /// <summary>
    /// Mở listener UDP local và bắt đầu link TCP.
    /// Trả về cổng UDP local mà WireGuard phải trỏ tới.
    /// </summary>
    public ushort Start()
    {
        Socket udp;
        try
        {
            udp = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
            udp.Bind(new IPEndPoint(IPAddress.Loopback, 0));
        }
        catch (SocketException ex)
        {
            throw new WireGuardConfigException($"relay socket error: {ex.Message}");
        }

        LocalPort = (ushort)((IPEndPoint)udp.LocalEndPoint!).Port;

        lock (_lock)
        {
            _udp = udp;
            _running = true;
        }

        _readTask = Task.Run(ReadLoopAsync);
        _writeTask = Task.Run(WriteLoop);
        _heartbeatTask = Task.Run(HeartbeatLoopAsync);

        _log.Info($"relay: local udp listener 127.0.0.1:{LocalPort} -> {_host}:[{string.Join(",", _ports)}]");
        return LocalPort;
    }

    public void Stop()
    {
        lock (_lock)
        {
            if (!_running)
            {
                return;
            }
            _running = false;
        }

        _cts.Cancel();

        Socket? tcp;
        Socket? udp;
        lock (_lock)
        {
            tcp = _tcp;
            _tcp = null;
            udp = _udp;
            _udp = null;
        }

        CloseQuietly(tcp);
        CloseQuietly(udp);
        _log.Info("relay: stopped");
    }

    public string CountersSummary()
    {
        lock (_lock)
        {
            var tcpFd = _tcp is null ? -1 : _tcp.Handle.ToInt64();
            return $"udpFrames={_sentFrames} udpBytes={_sentBytes} " +
                   $"framesFromRelay={_receivedFrames} bytesFromRelay={_receivedBytes} " +
                   $"droppedNoLink={_udpDroppedNoLink} tcpFd={tcpFd}";
        }
    }

    // MARK: - TCP link

    /// <summary>Kết nối tới relay, thử lại chừng nào tunnel còn bật.</summary>
    private async Task<Socket?> ConnectTcpAsync(CancellationToken token)
    {
        while (IsRunning)
        {
            Socket socket;
            try
            {
                socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            }
            catch (SocketException ex)
            {
                _log.Error($"relay: tcp socket error: {ex.Message}");
                return null;
            }

            ushort port;
            lock (_lock)
            {
                port = _lastGoodPort ?? _ports[_portIndex % _ports.Length];
                if (_lastGoodPort is null)
                {
                    _portIndex++;
                }
            }

            IPAddress? address;
            try
            {
                address = await ResolveAsync(_host, token).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                _log.Error($"relay: cannot resolve {_host}: {ex.Message}");
                socket.Dispose();
                await DelayAsync(RetryDelayMs, token).ConfigureAwait(false);
                continue;
            }

            if (address is null)
            {
                _log.Error($"relay: cannot resolve {_host}");
                socket.Dispose();
                await DelayAsync(RetryDelayMs, token).ConfigureAwait(false);
                continue;
            }

            var started = Stopwatch.StartNew();
            var connected = false;
            var failureReason = string.Empty;
            try
            {
                using var connectTimeout = CancellationTokenSource.CreateLinkedTokenSource(token);
                connectTimeout.CancelAfter(ConnectTimeoutMs);
                await socket.ConnectAsync(new IPEndPoint(address, port), connectTimeout.Token).ConfigureAwait(false);
                connected = true;
            }
            catch (OperationCanceledException) when (!token.IsCancellationRequested)
            {
                // Lý do hỏng ghi RÕ, không dùng lại errno: đây là hết hạn 10s chờ
                // connect hoàn tất, không phải lỗi mạng cụ thể (bài học ở
                // WGRelayClient.swift:168-189).
                failureReason = $"hết hạn {ConnectTimeoutMs / 1000}s chờ connect hoàn tất";
            }
            catch (Exception ex)
            {
                failureReason = ex.Message;
            }

            if (!connected)
            {
                _log.Info($"relay: connect {_host}:{port} failed: {failureReason}");
                socket.Dispose();
                await DelayAsync(RetryDelayMs, token).ConfigureAwait(false);
                continue;
            }

            try
            {
                socket.NoDelay = true;
            }
            catch (SocketException)
            {
                // Không bật được TCP_NODELAY không phải lỗi chí mạng.
            }

            lock (_lock)
            {
                _lastGoodPort = port;
            }
            _log.Info($"relay: connected to {_host}:{port} in {started.ElapsedMilliseconds}ms");
            return socket;
        }

        return null;
    }

    /// <summary>Đọc frame độ dài tiền tố và giao từng payload cho WireGuard dưới dạng datagram.</summary>
    private async Task ReadLoopAsync()
    {
        var token = _cts.Token;
        while (IsRunning)
        {
            var socket = await ConnectTcpAsync(token).ConfigureAwait(false);
            if (socket is null)
            {
                return;
            }

            lock (_lock)
            {
                _tcp = socket;
            }

            var header = new byte[2];
            while (IsRunning)
            {
                if (!ReadExactly(socket, header, 2))
                {
                    break;
                }
                var frameLength = (header[0] << 8) | header[1];
                if (frameLength <= 0 || frameLength > MaxDatagramBytes)
                {
                    continue;
                }
                var payload = new byte[frameLength];
                if (!ReadExactly(socket, payload, frameLength))
                {
                    break;
                }
                lock (_lock)
                {
                    _receivedFrames++;
                    _receivedBytes += frameLength;
                }
                SendToWireGuard(payload);
            }

            lock (_lock)
            {
                if (ReferenceEquals(_tcp, socket))
                {
                    _tcp = null;
                }
            }
            CloseQuietly(socket);
            if (IsRunning)
            {
                _log.Info("relay: TCP link dropped, reconnecting");
            }
        }
    }

    /// <summary>Chuyển mọi datagram WireGuard vào link TCP.</summary>
    private void WriteLoop()
    {
        var buffer = new byte[MaxDatagramBytes];
        while (IsRunning)
        {
            Socket? udp;
            lock (_lock)
            {
                udp = _udp;
            }
            if (udp is null)
            {
                Thread.Sleep(200);
                continue;
            }

            EndPoint sender = new IPEndPoint(IPAddress.Any, 0);
            int received;
            try
            {
                received = udp.ReceiveFrom(buffer, ref sender);
            }
            catch (ObjectDisposedException)
            {
                return;
            }
            catch (SocketException)
            {
                if (IsRunning)
                {
                    Thread.Sleep(200);
                }
                continue;
            }

            if (received <= 0)
            {
                continue;
            }

            Socket? tcp;
            lock (_lock)
            {
                _peer = (IPEndPoint)sender;
                tcp = _tcp;
            }

            // Chưa có link: bỏ đi — WireGuard tự gửi lại handshake mỗi 5s.
            if (tcp is null)
            {
                lock (_lock)
                {
                    _udpDroppedNoLink++;
                }
                continue;
            }

            var frame = new byte[received + 2];
            frame[0] = (byte)((received >> 8) & 0xff);
            frame[1] = (byte)(received & 0xff);
            Buffer.BlockCopy(buffer, 0, frame, 2, received);

            lock (_lock)
            {
                _sentFrames++;
                _sentBytes += received;
            }

            if (!WriteAll(tcp, frame))
            {
                // Link hỏng một chiều (NAT khách sạn / relay restart): phá link để
                // read loop reconnect. Không làm vậy thì mọi gói sau bị bỏ âm thầm
                // và tunnel "up" mà không có traffic (WGRelayClient.swift:276-282).
                _log.Info("relay: write to relay failed — dropping link to reconnect");
                DropLink(tcp);
            }
        }
    }

    // MARK: - Socket helpers

    /// <summary>Đóng link hỏng; read loop sẽ nhận ra và reconnect.</summary>
    private void DropLink(Socket socket)
    {
        _log.Info($"relay: dropping TCP link: write failure");
        lock (_lock)
        {
            if (ReferenceEquals(_tcp, socket))
            {
                _tcp = null;
            }
        }
        CloseQuietly(socket);
    }

    private void SendToWireGuard(byte[] payload)
    {
        Socket? udp;
        IPEndPoint? target;
        lock (_lock)
        {
            udp = _udp;
            target = _peer;
        }
        if (udp is null || target is null)
        {
            return;
        }
        try
        {
            udp.SendTo(payload, target);
        }
        catch (SocketException)
        {
            // Best effort.
        }
        catch (ObjectDisposedException)
        {
            // Đang Stop().
        }
    }

    private static bool ReadExactly(Socket socket, byte[] buffer, int count)
    {
        var offset = 0;
        while (offset < count)
        {
            int read;
            try
            {
                read = socket.Receive(buffer, offset, count - offset, SocketFlags.None);
            }
            catch (SocketException)
            {
                return false;
            }
            catch (ObjectDisposedException)
            {
                return false;
            }

            if (read > 0)
            {
                offset += read;
            }
            else if (read == 0)
            {
                return false; // peer closed
            }
        }
        return true;
    }

    private static bool WriteAll(Socket socket, byte[] bytes)
    {
        var offset = 0;
        while (offset < bytes.Length)
        {
            int written;
            try
            {
                written = socket.Send(bytes, offset, bytes.Length - offset, SocketFlags.None);
            }
            catch (SocketException)
            {
                return false;
            }
            catch (ObjectDisposedException)
            {
                return false;
            }

            if (written > 0)
            {
                offset += written;
            }
            else
            {
                return false;
            }
        }
        return true;
    }

    private static async Task<IPAddress?> ResolveAsync(string host, CancellationToken token)
    {
        if (IPAddress.TryParse(host, out var literal))
        {
            return literal.AddressFamily == AddressFamily.InterNetwork ? literal : null;
        }

        var addresses = await Dns.GetHostAddressesAsync(host).ConfigureAwait(false);
        foreach (var address in addresses)
        {
            if (address.AddressFamily == AddressFamily.InterNetwork)
            {
                return address;
            }
        }
        return null;
    }

    private static async Task DelayAsync(int milliseconds, CancellationToken token)
    {
        try
        {
            await Task.Delay(milliseconds, token).ConfigureAwait(false);
        }
        catch (OperationCanceledException)
        {
            // Stop().
        }
    }

    private async Task HeartbeatLoopAsync()
    {
        var token = _cts.Token;
        try
        {
            while (IsRunning)
            {
                await Task.Delay(TimeSpan.FromSeconds(HeartbeatSeconds), token).ConfigureAwait(false);
                if (!IsRunning)
                {
                    return;
                }
                _log.Info($"relay: heartbeat {CountersSummary()}");
            }
        }
        catch (OperationCanceledException)
        {
            // Stop().
        }
    }

    private bool IsRunning
    {
        get
        {
            lock (_lock)
            {
                return _running;
            }
        }
    }

    private static void CloseQuietly(Socket? socket)
    {
        if (socket is null)
        {
            return;
        }
        try
        {
            socket.Shutdown(SocketShutdown.Both);
        }
        catch (Exception)
        {
            // Socket có thể chưa connect; bỏ qua.
        }
        try
        {
            socket.Close();
        }
        catch (Exception)
        {
            // Best effort.
        }
    }

    public void Dispose()
    {
        Stop();
        _cts.Dispose();
    }
}
